platform/android/src/jni.cpp

```
@@ -1620,6 +1753,363 @@ void JNICALL nativeRemoveCustomLayer(JNIEnv *env, jobject obj, jlong nativeMapVi
+    mbgl::DefaultFileSource *defaultFileSource = reinterpret_cast<mbgl::DefaultFileSource *>(defaultFileSourcePtr);
+    defaultFileSource->listOfflineRegions([defaultFileSourcePtr, listCallback](std::exception_ptr error, mbgl::optional<std::vector<mbgl::OfflineRegion>> regions) {
+
+        // Reattach, the callback comes from a different thread
+        JNIEnv *env2;
+        jboolean renderDetach = attach_jni_thread(theJVM, &env2, "Offline Thread");
+        if (renderDetach) {
+            mbgl::Log::Debug(mbgl::Event::JNI, "Attached.");
+        }
+
+        if (error) {
+            std::string message = mbgl::util::toString(error);
+            env2->CallVoidMethod(listCallback, listOnErrorMethodId, std_string_to_jstring(env2, message));
+        } else if (regions) {
+            // Build the OfflineManager object
+            jobject jmanager = env2->NewObject(offlineManagerClass, offlineManagerClassConstructorId);
+            env2->SetLongField(jmanager, offlineManagerClassPtrId, defaultFileSourcePtr);
+
+            // Build jobjectArray
+            jsize index = 0;
+            jobjectArray jregions = env2->NewObjectArray(regions->size(), offlineRegionClass, NULL);
+            for (std::vector<mbgl::OfflineRegion>::iterator region = regions->begin(); region != regions->end(); ++region) {
+                // Build the Region object
+                jobject jregion = env2->NewObject(offlineRegionClass, offlineRegionConstructorId);
+                env2->SetObjectField(jregion, offlineRegionOfflineManagerId, jmanager);
+                env2->SetLongField(jregion, offlineRegionIdId, region->getID());
+                env2->SetObjectField(jregion, offlineRegionMetadataId, metadata_from_native(env2, region->getMetadata()));
+
+                // Moves the region on the stack into a heap-allocated one
+                env2->SetLongField(jregion, offlineRegionPtrId,
+                    reinterpret_cast<jlong>(new mbgl::OfflineRegion(std::move(*region))));
```

Another wrinkle with the above is that `OfflineRegion.delete(OfflineRegionDeleteCallback)` puts the region into a state where it is invalid to call further methods on it, because the native side releases ownership of the `mbgl::OfflineRegion` ([here](https://github.com/mapbox/mapbox-gl-native/blob/f67efbf701829ae7ab121e4f7bcfcf7f497de4bc/platform/android/src/jni.cpp#L2085)). This has a couple of implications:

* `deleteOfflineRegion` (the native method) should reset the value of the `mOfflineRegionPtr` field to 0.
* Other `OfflineRegion` methods should raise an exception if the value of `mOfflineRegionPtr` is 0, in case someone tries to call the method after calling `OfflineRegion.delete`.